src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,178 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            Node<T> first = new Node<>(null, value, null);
+            head = first;
+            tail = first;
+            first.prev = null;
+            first.next = null;
+            size++;
+        } else {
+            Node<T> last = tail;
+            Node<T> newNode = new Node<>(last, value, null);
+            newNode.prev = tail;
+            last.next = newNode;
+            tail = newNode;
+            size++;
+        }
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (size == 0) {
+            Node<T> first = new Node<>(null, value, null);
+            head = first;
+            tail = first;
+            first.prev = null;
+            first.next = null;
+            size++;
+        } else if (index == size) {
+            Node<T> last = tail;
+            Node<T> newNode = new Node<>(last, value, null);
+            newNode.prev = tail;
+            last.next = newNode;
+            tail = newNode;
+            size++;
+        } else {
+            Node<T> current = node(index);
+            Node<T> left = current.prev;
+            Node<T> newNode = new Node<>(left, value, current);
+            if (left == null) {
+                head = newNode;
+            } else {
+                current.prev = newNode;
+                left.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexChecker(index);
+        return node(index).items;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexChecker(index);
+        Node<T> setElement = head;
+        for (int i = 0; i < index; i++) {
+            setElement = head.next;
+        }
+        T oldValue = setElement.items;
+        setElement.items = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexChecker(index);
+        Node<T> oldNode = node(index);
+        Node<T> prevNode = node(index).prev;
+        Node<T> nextNode = node(index).next;
+        final T returnElement = oldNode.items;
+        if (oldNode.prev == null) {
```

Comment:
```suggestion
        if (prevNode == null) {
```
use the same variables to make your code more readable